### ISSUE TITLE
Question about Subscribe-Unsubscribe in Streams and overal streams approach

### ISSUE THREAD

**Issue by centur** [dotnet/orleans#1272]
I have a bit of conceptual question about Streams and Implicit subscription (cc @gabikliot ) :

When some code produces an event to a stream - this triggers implicit subscription for all interested listeners. 
Implicit subscription uses pub-sub grain which creates corresponding subscription record in storage table (with RowKey: "Orleans.Streams.PubSubRendezvousGrain" ), and the record's Id is based on StreamId +StreamProvider+Namespace like this:

```
00000000-0000-0000-0000-000000000000_GrainReference=0000000000000000000000000000000006000000716e8e94+GrainCommands_PermissionGrainDelete
00000000-0000-0000-0000-000000000000_GrainReference=40aa265e822d8647b183fe7169b2dea806000000716e8e94+GrainCommands_PermissionGrainDelete
00000000-0000-0000-0000-000000000000_GrainReference=43dd2abc75fce92d4e45088da281c6b206000000716e8e94+GrainCommands_PermissionGrainDelete
00000000-0000-0000-0000-000000000000_GrainReference=4812485b248eee8ed920ff4983b0caa706000000716e8e94+GrainCommands_PermissionGrainDelete
```

Question 1: if one is using Guid.NewId() for every message produced - does this mean that eventually it'll fill the database with subscription records that were used only once and then become obsolete and will be never read again (unless some other producer will generate exactly same guid with very low chances to do this) ? Is this correct ? 

Should the listener grain code at least unsubscribe from stream in ```OnDeactivateAsync()``` to cleanup after himself, otherwise it'll bloat the storage with no good reason?

Also this approach initiates at least one Storage-Write, Storage-Read and grain activation for each message produced, not talking about other in-code un-necessities. This part looks like a flaw of Streams design.

Question 2: using Guid.Empty as stream Id creates extremely hot grains - as it's exactly single grain to handle all messages in the namespace (or multiple namespaces). I can use some tricks like converting fixed Int to guids (like described here http://stackoverflow.com/questions/4825907/convert-int-to-guid)  but it feels that it's not the right way to spread the load. What practices are you guys using for spreading the load but keep it sane and without hardcoding guids?

**Comment by gabikliot** [comment]
Implicit subscriptions do NOT uses pub-sub and do NOT create corresponding subscription records in storage table.
Matching producer with implicit subscribers is completely stateless (based on static code analysis of the attributes), and does not store any dynamic data in any table.

Explicit subscriptions DO use pub-sub and do create corresponding subscription record in storage table.

I will answer your questions assuming you meant to ask about **Explicit subscriptions**:
1) Think about streams just like grains. Of course you can create a new grain/stream for every operation/message, but each one does have a cost. So just like with actors we suggest to have enough to spread the load but not too little to create bottleneck - same is true with streams. Don't create a stream per msg. Create enough, but not too much streams, to spread the load.

>Should the listener grain code at least unsubscribe from stream in OnDeactivateAsync() to cleanup after himself, otherwise it'll bloat the storage with no good reason?

Yes.

>Storage-Write, Storage-Read and grain activation for each message produced, not talking about other in-code un-necessities. This part looks like a flaw of Streams design.

Not a flaw, but a potential inefficiency, **in certain workloads**. To answer that:
a) we did have ideas to optimize pub sub more.
b) implicit pub sub is one such idea. It optimizes it completely - zero cost for pub sub.
c) before investing too much into this optimization, I would like to see **a realistic workload**, that could not be easily optimized otherwise. As of now, I am still to find this workload. Whenever we had a pub sub as a bottleneck, we discovered that we could use implicit instead. That is not to say that in the future there won't be a need for a more efficient pub sub. And we have ideas how to achieve that. 

2)  Not sure what you meant about Guids. Same rules as with Grains - come up with a naming strategy to spread the load between grains/streams.

**Comment by centur** [comment]
Quick note ( will add more later when get home) - from what I observing - Implicit subscription creates pubsub Grain state because as doc says [here](http://dotnet.github.io/orleans/Orleans-Streams/Streams-Programming-APIs)

>The presence of ImplicitStreamSubscription causes the streaming runtime to automatically subscribe this grain to a stream and deliver the stream events to it. However, the grain code still needs to tell the runtime how it wants events to be processed. Essentially, it need to attach the IAsyncObserver. Therefore, when the grain is activated, the grain code inside OnActivateAsync needs to call:
```
IStreamProvider streamProvider = base.GetStreamProvider("SimpleStreamProvider"); 
IAsyncStream<T> stream = streamProvider.GetStream<T>(this.GetPrimaryKey(), "MyStreamNamespace"); 
StreamSubscriptionHandle<T> subscription = await stream.SubscribeAsync(IAsyncObserver<T>);  
```

Just following this advice I managed to get implicit streams working. And I was under impression that ImplicitStreamSubscription is just a way to activate specific grain and let it subscribe to the stream BEFORE the message actually sent to subscribers.

Following these docs I can clearly see that it writes some State for ```Orleans.Streams.PubSubRendezvousGrain``` in the storage.

I've also have this provider in our config file
```
<Provider Name="PubSubStore" Type="Orleans.Storage.AzureTableStorage" DataConnectionString="snip!" />
```
which presumably defies that zero cost pub sub. 
Do you refer to zero cost only for ```Orleans.Storage.MemoryStorage``` based pub-sub ?

**Comment by gabikliot** [comment]
Where did you see that "Implicit subscription creates pubsub Grain state "?
It does not.

>Do you refer to zero cost only for Orleans.Storage.MemoryStorage based pub-sub

No. Implicit subscriptions do NOT write anything to any storage provider.

**Comment by gabikliot** [comment]
@centur, my bad: you MUST configure StreamPubSubType.ImplicitOnly
Otherwise, it does create an empty row in the pub sub table (since the same stream may have both implicit and explicit) subscribers. see here https://github.com/dotnet/orleans/blob/master/src/Tester/StreamingTests/ControllableStreamGeneratorProviderTests.cs#L54 how to configure that.
Once configured, you should see no updates to the pub sub table.

**Comment by centur** [comment]
:+1:  Thanks for detailed explanation, ```ImplicitOnly``` makes more sense.

Does declarative attribute in config also suppose to work ? 
If yes - I still can observe unexplained behaviour.

I have this in my config (both in Client and Silo config), as per your example
```
<Provider Name="GrainCommands" Type="Orleans.Providers.Streams.SimpleMessageStream.SimpleMessageStreamProvider"
					  PubSubType="ImplicitOnly" />
```

GrainCommands is an ImplicitOnly stream (actually all our streams are - we found little use for explicit streams so far). But still can see that something creates subscription.

I've flipped UseJsonFormat="true" on PubSubStore provider and can see this in the StringData field:
```
{
	"$type": "System.Collections.Generic.Dictionary`2[[System.String, mscorlib],[System.Object, mscorlib]], mscorlib",
	"Producers": {
		"$type": "System.Collections.Generic.HashSet`1[[Orleans.Streams.PubSubPublisherState, Orleans]], System.Core",
		"$values": [{
			"$type": "Orleans.Streams.PubSubPublisherState, Orleans",
			"Stream": {
				"$type": "Orleans.Streams.StreamId, Orleans",
				"Guid": "00000000-0000-0000-0000-000000000000",
				"ProviderName": "GrainCommands",
				"Namespace": "PermissionGrainDelete"
			},
			"producerReference": {
				"$type": "Orleans.Streams.OrleansCodeGenStreamProducerExtensionReference, Orleans",
				"GrainId": "0000000000000000000000000000000006ffffffa06a4a58+5ed2183f-949c-40d4-b9f4-98ae9c6d8fdd",
				"GenericArguments": ""
			},
			"Producer": {
				"$type": "Orleans.Streams.OrleansCodeGenStreamProducerExtensionReference, Orleans",
				"GrainId": "0000000000000000000000000000000006ffffffa06a4a58+5ed2183f-949c-40d4-b9f4-98ae9c6d8fdd",
				"GenericArguments": ""
			}
		}]
	},
	"Consumers": {
		"$type": "System.Collections.Generic.HashSet`1[[Orleans.Streams.PubSubSubscriptionState, Orleans]], System.Core",
		"$values": []
	}
}
```

So there is a bug - something explicitly subscribes to an ImplicitOnly stream OR provider configuration doesn't respect that attribute in xml config.

**Comment by centur** [comment]
regarding other questions and your answers - I think questions will be pretty much obsolete once we will come to the common ground  how to properly use implicit streams which comes with zero cost subscription.

**Comment by gabikliot** [comment]
Try `StreamPubSubType.ExplicitGrainBasedAndImplicit`.
Also, just look at the log - when provider is loaded it prints what it uses for configuration.

**Comment by gabikliot** [comment]
@centur , did you try `StreamPubSubType.ExplicitGrainBasedAndImplicit`?

**Comment by centur** [comment]
Yes, just tried it and combination of providers:

```AzureQueueStreamProvider``` - does respect value from ```PubSubType``` attribute and when switched to ImplicitOnly - does not write any data to PubSubStore.
Interesting though it doesn't write anything if the value is ```ExplicitGrainBasedAndImplicit```, but it might be because I don't have anything subscribing to it explicitly

```SimpleMessageStreamProvider```- doesn't respect ```PubSubType``` value from config and always creates PubSubGrain State.  Because my PubSubStore configured as ```Name="PubSubStore" Type="Orleans.Storage.AzureTableStorage"``` - the record with some state appears in there.

**Comment by gabikliot** [comment]
@centur ,`ExplicitGrainBasedAndImplicit` is the default, so it will be just like when nothing is specified. Therefore, with it, even if you do not have explicit streams, it will still write producers to the table.

As for SMS (`SimpleMessageStreamProvider`), indeed, we forgot to add pub sub config to it. I will fix it.

**Comment by centur** [comment]
@gabikliot Umm your last comment describes an opposite behaviour to what I'm seeing with Explicit+Implicit.
 This is what I saw
>Interesting though it **DOES NOT** write anything if the value is **ExplicitGrainBasedAndImplicit**, but it might be because I don't have anything subscribing to it explicitly

This is what you said it should be

>Therefore, with it, even if you do not have explicit streams, it **will still write** producers to the table

Is there a typo in your answer or it might be another bug around configured and actual behaviour ?

**Comment by gabikliot** [comment]
No, I still stand behind what I wrote.
`ExplicitGrainBasedAndImplicit` is the default configuration. It should behave the same way as if no pub sub type was defined (which is the setup you described at the beginning of this issue).

If you see that `ExplicitGrainBasedAndImplicit` behaves differently than default (no configuration), please submit a test with this setup as a repro and we will look into that.

**Comment by centur** [comment]
No. default is "Explicit+Implicit" and there is no difference in behaviour when PubSubType attr is not specified and when it's explicitly set to ```ExplicitGrainBasedAndImplicit ``` . But you mentioned that with 
```ExplicitGrainBasedAndImplicit ``` it will write producers to the table
Did I understand this correct and you're saying that even if system doesn't have Explicit streams -  it will register producers on PubSubGrain and this (my PubSubStore is AzureStorage based) must add pub-sub state record to the table ? If this is correct - I don't see such behaviour (this is what I've mentioned in "Interesting..." in my comment above )

**Comment by gabikliot** [comment]
It is very simple: 
1) if you use Implicit Only, no table will be used. Guaranteed.
2) if you use Explicit+Implicit, or default nothing, table may be used. The system may write stuff into table. 

That is what one needs to know in terms of guarantees. If you don't want table to be used - use implicit. If you cannot use implicit, use regular pub sub and the table may/will be used.

Beyond that, it's implementation details not semantics.

**Comment by centur** [comment]
ok

**Comment by centur** [comment]
Umm, I need to double check it, but seems that even with proper PubSubType configuration :
```<Provider Name="GrainCommands" Type="Orleans.Providers.Streams.SimpleMessageStream.SimpleMessageStreamProvider" PubSubType="ImplicitOnly" />```
(log confirms that PubSubType is properly parsed and set to ImplicitOnly)

SimpleMessageStreamProvider creates PubSub record in PubSubStorage.

**Comment by gabikliot** [comment]
I think this is a wrong sintax. The pub sub type should be a full qualified enum including namespace. See above the exact syntax I wrote. If it doesn't behave like you think it should, please share a silo log.

**Comment by centur** [comment]
This is the right syntax. 
Based on the code in Orleans the value parsed like this
```PubSubType =  (StreamPubSubType)Enum.Parse(typeof(StreamPubSubType), pubSubTypeString);```	

Here is a linqpad snippet for the same code with full enum name:
```
// This is a linqpad snippet you can copy and run 
void Main()
{
    //fully qualified enum string doesn't work
	string pubSubTypeString = "StreamPubSubType.ImplicitOnly";
	var PubSubType =  (StreamPubSubType)Enum.Parse(typeof(StreamPubSubType), pubSubTypeString);
// Throws ArgumentException: Requested value 'StreamPubSubType.ImplicitOnly' was not found.
	PubSubType.Dump();
}
public enum StreamPubSubType
{
	ExplicitGrainBasedAndImplicit,
	ExplicitGrainBasedOnly,
	ImplicitOnly,
}
```

Same code with just a member name works perfectly
```
// This is a linqpad snippet you can paste in and run 
void Main()
{
    //enum member string does work
	string pubSubTypeString = "ImplicitOnly";
	var PubSubType = (StreamPubSubType)Enum.Parse(typeof(StreamPubSubType), pubSubTypeString);
	PubSubType.Dump();
}
public enum StreamPubSubType
{
	ExplicitGrainBasedAndImplicit,
	ExplicitGrainBasedOnly,
	ImplicitOnly,
}

```
So the syntax is correct. Also with ```PubSubType="StreamPubSubType.ImplicitOnly"``` my local silo doesn't start, with ```PubSubType="ImplicitOnly"``` it does

Here is the silo log attached and it shows that SimpleMessageStreamProvider configured properly
[Silo_AzureCloudServiceHost.DBCloud.ActorsHost_IN_0-2016-01-28-10.28.34.277Z.txt](https://github.com/dotnet/orleans/files/108032/Silo_AzureCloudServiceHost.DBCloud.ActorsHost_IN_0-2016-01-28-10.28.34.277Z.txt)

>[2016-01-28 10:28:47.290 GMT    38	INFO   	0	SimpleMessageStreamProvider	127.255.0.0:11111]	Initialized SimpleMessageStreamProvider with name Notifications and with property FireAndForgetDelivery: False and PubSubType: ImplicitOnly	
[2016-01-28 10:28:47.293 GMT    38	INFO   	0	SimpleMessageStreamProvider	127.255.0.0:11111]	Initialized SimpleMessageStreamProvider with name CoreEvents and with property FireAndForgetDelivery: False and PubSubType: ImplicitOnly	
[2016-01-28 10:28:47.300 GMT    38	INFO   	0	SimpleMessageStreamProvider	127.255.0.0:11111]	Initialized SimpleMessageStreamProvider with name GrainCommands and with property FireAndForgetDelivery: False and PubSubType: ImplicitOnly	



And here what I get in OrleansGrainState table ( it's a single record there, I've cleaned up everything before running this test)

```
PartitionKey:00000000-0000-0000-0000-000000000000_GrainReference=4715ac67f92713d76e4bcf6ceefbdba106000000716e8e94+GrainCommands_PageAnnotationUpdate
RowKey:Orleans.Streams.PubSubRendezvousGrain
Etag:W/"datetime'2016-01-28T10%3A29%3A11.1057247Z'"
StringData:
{
  "$type": "System.Collections.Generic.Dictionary`2[[System.String, mscorlib],[System.Object, mscorlib]], mscorlib",
  "Producers": {
    "$type": "System.Collections.Generic.HashSet`1[[Orleans.Streams.PubSubPublisherState, Orleans]], System.Core",
    "$values": [
      {
        "$type": "Orleans.Streams.PubSubPublisherState, Orleans",
        "Stream": {
          "$type": "Orleans.Streams.StreamId, Orleans",
          "Guid": "f92713d7-ac67-4715-a1db-fbee6ccf4b6e",
          "ProviderName": "GrainCommands",
          "Namespace": "PageAnnotationUpdate"
        },
        "producerReference": {
          "$type": "Orleans.Streams.OrleansCodeGenStreamProducerExtensionReference, Orleans",
          "GrainId": "4a98943c164bc00f291e562e64bd46950400000000000000",
          "ObserverId": "d3d6839b-bb1d-4787-9563-9efe9e0e3d0a",
          "GenericArguments": ""
        },
        "Producer": {
          "$type": "Orleans.Streams.OrleansCodeGenStreamProducerExtensionReference, Orleans",
          "GrainId": "4a98943c164bc00f291e562e64bd46950400000000000000",
          "ObserverId": "d3d6839b-bb1d-4787-9563-9efe9e0e3d0a",
          "GenericArguments": ""
        }
      }
    ]
  },
  "Consumers": {
    "$type": "System.Collections.Generic.HashSet`1[[Orleans.Streams.PubSubSubscriptionState, Orleans]], System.Core",
    "$values": []
  }
}
```

To trigger that i have a special debug controller which is doing this:
```
			var streamProvider = GrainClient.GetStreamProvider(StreamNames.GrainCommands_Provider);
			var producer = streamProvider.GetStream<GrainCommandData>(streamId, nameof(GrainCommandType.PageAnnotationUpdate));
			await producer.OnNextAsync(new GrainCommandData { EventType = GrainCommandType.PageAnnotationUpdate});
```

The onle place I subscribe to "PageAnnotationUpdate" namespace is my grain with these attributes
```
[ImplicitStreamSubscription(nameof(GrainCommandType.PageAnnotationUpdate))]
	[ImplicitStreamSubscription(nameof(GrainCommandType.PageAnnotationDelete))]
	public class PageAnnotationsListener: BaseListenerGrain<GrainCommandData>, IPageAnnotationsListener
```

I'm really sorry for being pain in the ass with this issue, but there is something writes to pubsub for SimpleMessageProvider and ImplicitOnly stream subscription.

**Comment by gabikliot** [comment]
>I'm really sorry for being pain in the ass with this issue, but there is something writes to pubsub for SimpleMessageProvider and ImplicitOnly stream subscription.

On the contrary! Thank you for pushing and not giving up on this!

With your log, I was now been able to validate that there is a bug. Your configuration is correct, yet we still for some reason activate `PubSubRendezvousGrain` (which will store its state in Azure Table) (the fact that we do activate it is evident from WARNING 101215	at the bottom of the log ). With  `ImplicitOnly` we should not be creating  PubSubRendezvousGrains at all (for that stream provider).
So it is a bug and we will look into it.

I wander if @jason-bragg has some idea?

**Comment by sergeybykov** [comment]
@centur +1 to what @gabikliot said. This is not pain, this is a valuable contribution - to find and articulate a scenario that doesn't work as expected.

Should we open a separate issue for this?

**Comment by centur** [comment]
@gabikliot @sergeybykov  ok I'll continue then :). I think this is the issue where this started while ago as a question around zero cost implicit subscription and it surfaced, at first, attribute parsing issues and now this one. If you feel that it's better to create a new issue - IMO it worth to reference back to this as it has the discussion and context.

**Comment by gabikliot** [comment]
I think its fine to leave it here, with all the right context.

**Comment by jason-bragg** [comment]
I'm spending some time today looking into test failure, so I'll look for this as I'm poking around.

@centur, Maybe you already stated this, and I missed it somewhere in the thread, but what version of Orleans are you using?

**Comment by centur** [comment]
@jason-bragg we are using latest nuget release, 1.1.2

We started to use streams a lot around 1.1 and I've raised this question  from 1.1 (or 1.1.1) and thought that it was fixed in 1.1.2 as #1285 PR was merged. I haven't checked all cases as had to focus on other things. Quick check confirmed what @gabikliot said - it is zero cost for AzureQueue stream provider and properly configured pubsub type (ImplicitOnly).

**Comment by jason-bragg** [comment]
@centur 
I apologize for the delayed reply.  I've looked into this and my tests running sms configured to use ImplicitOnly does not touch storage.

Can you confirm that you've sms configured to use ImplicitOnly both on the silo and client configurations.

**Comment by jason-bragg** [comment]
this is my stream provider configuration code:

       {
                ...
                config.Globals.RegisterStreamProvider<SimpleMessageStreamProvider>(StreamProviderName, BuildProviderSettings())
                ...
        }

        private static Dictionary<string, string> BuildProviderSettings()
        {
            return new Dictionary<string, string>
            {
                // add pub/sub settting
                {"PubSubType", StreamPubSubType.ImplicitOnly.ToString()}
            };
        }

**Comment by jason-bragg** [comment]
Is there any change that something is trying to use the GrainCommands stream provider before it is fully initialized?

**Comment by centur** [comment]
>Can you confirm that you've sms configured to use ImplicitOnly both on the silo and client configurations.

Umm it's not, I pay attention only to silo configuration. Do I need to reflect all settings in client configuration too ?

PS: Sorry I haven't had a chance to get back to all pending issues earlier due to other duties..

**Comment by jason-bragg** [comment]
>Do I need to reflect all settings in client configuration too ?

Both the client and silo stream providers use the pubsub system, so if either is configured to something other than ImplicitOnly, then pubsub grains could be activated.

If all stream providers on client and server are all configured to use ImplicitOnly and you're still seeing pubsub grain state being written to storage, try a run with the "PubSubStore" storage provider removed from the silo configuration.  This should cause errors that may provide more context to when/where the wrong pubsub is being called.  If you try this, please send logs, and i'll take another look.

>due to other duties..

No worries.  I'm in the same boat :/

**Comment by gabikliot** [comment]
Hmmm, I thought you wrote above: 

>I have this in my config (both in Client and Silo config), as per your example
```
<Provider Name="GrainCommands" Type="Orleans.Providers.Streams.SimpleMessageStream.SimpleMessageStreamProvider"
					  PubSubType="ImplicitOnly" />
```

**Comment by centur** [comment]
@gabikliot Previously I have to set up and almost mirror stream providers section for both silo and client, but IIRC I found that I just need to keep names\types and conn-strings in sync. I haven't thought that I need to keep all attributes on these in sync and was thinking that I just need to preserve key elements, and things like PubSubType are playing only on server side to define how stream dispatcher (silo) stores it's own "system-related" settings

**Comment by gabikliot** [comment]
OK, great, lets configure the client correctly and validate that it indeed behaves like we expect it to.

**Comment by centur** [comment]
Configured all stream providers on Silo AND Client with PubSubType=ImplicitOnly.
Can confirm - no PubSub record being added. 
Removed PubSubType from client's configuration - record appears again.

Cause found, misconfiguration issue.
Am I correct assuming that ```	<StreamProviders>``` section in Silo and in Client configs must be 100% synchronized with each other ?

**Comment by gabikliot** [comment]
Yes, and not just stream providers - everything has to match.

**Comment by sergeybykov** [comment]
Mystery solved.

**Comment by gabikliot** [comment]
:+1:
